src/predict_api/flask_app.py

```diff
@@ -0,0 +1,243 @@
+import json
+
+import requests
+from apispec import APISpec
+from apispec_webframeworks.flask import FlaskPlugin
+from apispec.ext.marshmallow import MarshmallowPlugin
+from dotenv import load_dotenv
+from flask import Flask, g, redirect, request, url_for, jsonify
+from flask_login import (
+    LoginManager,
+    login_required,
+    login_user,
+    logout_user,
+)
+from flask_cors import CORS
+from flask_sqlalchemy import SQLAlchemy
+from oauthlib.oauth2 import WebApplicationClient
+
+from config import Config, logging
+
+from flask import jsonify, render_template, Blueprint
+from flask_login import login_required
+
+from config import logging, Config
+
+
+
+# TODO: many below
+# figure out topology. This should run in a separate container. Are requests all proxied through the annotation api?
+# auth - either it runs behind the main container or we need separate auth
+# CORS lockdown - we're currently allowing any cross-origin request
+# options to add to the API:
+#   - return diagnostic on processing time
+#   - versions of the inference algo - over time we want this more robust
+# model versioning.  Maybe the model comes from a blob storage url?
+# data model for image upload -> predictions -> annotations
```

Comment:
Close to this.  I'm thinking about the annotation experience when the image contains 10+ headstamps - it makes sense for the annotation flow to cycle through the individual predictions as separate sub-images.  Enable the bulk annotation experience to focus on one headstamp at a time.

Reply:
yes, we have been ignoring this for a while. One image can have many headstamps...
```
image_collection
   |
  /|\
image -- (fingerprint)
   |
  /|\
headstamp -- (coordinates)
   |
  /|\
annotation -- (attributes)
  \|/
   |
agent
```
It will be a bit of a refactor when it lands :/

Reply:
we can be strict and just say "one image always have one headstamp", adding another layer of NN which splits image with many headstamps to multiple images. Extra complexity but it's the way to do things right

Reply:
I don't think we can get away with saying that forever, because:
* someone may take a high-resolution picture of a scene that has multiple headstamps in it. Nice that the "where's Wally?" headstamp detector finds them for the user
* someone may rack up a box of cartridges on a flatbed scanner and scan them all at once.

Maybe we can have two kinds of images:
* "raw" multi-headstamp images (with zero or one headstamps)
* "clean" single headstamp images (with exacly one headstamp)

And then we could make clean images from raw ones, either/both with manual "draw a rectangle around the headstamp" workflow and with an automated version of the same (it kind of works already :). But if we do it, why is it really any different to recording a "headstamp" as a set of coordinates in an image - I mean, that's exactly what you would do to slice up the raw image anyway. So saving the "clean" image is really just a cache of slicing the coordinate.

Reply:
my point exactly is in user uploading whatever they want but some (micro)service parsing the image saying "I'll extract all headstamps from that image, stripping everything which is grass, background and ceiling" - returning from 0 to N resulting images

I see it being useful for marking up but probably a bit too complex for a prototype

Reply:
OK, I think we are talking about the same thing. Rob's image above (with multiple with the red circles) is example output of a model that says "these are the coordinates of the headstamps in the image".  We can save the coordinates and crop the images on demand (client side or server side), or we can crop the image once and save the output single-headstamp images.

I prefer to think about this as "save the coordinates" because we might have people training the headstamp detector. That's not a headstamp (it's a coin, or a pencil, or the moon). You missed this headstamp here, etc. So the set of coordinates is like a set of headstamp glyph annotations, it's something that's maintained with effort so the model performance can improve.

If we had an API that was basically `GET /collections/{collection_id}/{image_id}/{headstamp_id}` then the API could dynamically crop the sub-image but we could hide behind a CDN so it only does it once (or very infrequently). Just-in-time. So the SPA could pull them through the CDN (expensive only once). Or maybe it could even be done client side?